### DISCUSSION TITLE
Unexpected behaviour with ORM and `union_all`

### DISCUSSION THREAD

**Discussion by balint-backmaker** [sqlalchemy/sqlalchemy#9531] [Usage Questions] [answered]
## Behaviour description
In the code below...
- I would expect `session.scalars(stmt).all()` to return `Sequence[RawEvent]` as per the [ORM integration in the docs](https://docs.sqlalchemy.org/en/20/tutorial/data_select.html#selecting-orm-entities-from-unions)
- But in fact it returns the first column of `raw_event` only - this surprised me since I started all `queries` with the mapped object
- This suggests that `union_all` breaks the ORM integration via returning not a tuple of objects, but the raw columns instead
- if I leave it as `session.execute(stmt).all()` then I get `Sequence[Row]` which makes sense, but then how do you retain ORM aware objects and/or cast to `RawEvent` without manual mapping?

This looks a lot like a bug, but it may be that this is not ever possible to make with `union_all` because of some complicated type system quirk.

Can you show me a sensible workaround or fix to my problem?

## Example code
```python
from sqlalchemy import select, GenerativeSelect
from sqlalchemy import union_all
from sqlalchemy.orm import Session

def last_for_reference(reference: str) -> GenerativeSelect | Any:
    return select(RawEvent) \
        .where(RawEvent.reference == reference) \
        .order_by(RawEvent.received_at.desc()) \
        .limit(1)


def fetch_references(session: Session, references: list[str]) -> Sequence[RawEvent]:
    # here we try to avoid many return trips via a union_all
    queries = [last_for_reference(ref) for ref in references]
    stmt = union_all(*queries)
    logger.info(stmt) # this looks fine! ->  (SELECT ... LIMIT ...) UNION ALL (SELECT ... LIMIT ...)
    results = session.execute(stmt).all() # this returns low level rows, not high-level-tuples
    return results
```

## SQLAlchemy version
taken from `pip freeze` -> `SQLAlchemy==2.0.7`

**Comment by balint-backmaker** [comment, marked as answer]
muppet strikes -> I missed the literal next paragraph after the example in the docs
```
select(RawEvent).from_statement(union)
```

**Reply by CaselIT** [reply]
Yes, this is the expected behaviour of union

**Reply by balint-backmaker** [reply]
Thanks for the confirmation! :)
